### PR TITLE
Feature/uplink authorization

### PR DESCRIPTION
**Pull request by ramonornela**
**Type:** feature

**Description:**
Allowing to use NPM_TOKEN (or token assign in process.env) in uplink see http://blog.npmjs.org/post/118393368555/deploying-with-npm-private-modules in headers is also allowed, but it is better to assign in process . env much safer(allow make in CI)

### CODE REVIEW

**Comment by juanpicado** [comment]
The tests have to pass, please check them.

**Comment by ramonornela** [comment]
@juanpicado done

File: src/lib/storage/up-storage.js [hunk 1]
```diff
@@ -227,6 +227,57 @@ class ProxyStorage {
...
     // registry.npmjs.org will only return search result if user-agent include string 'npm'
     headers[userAgent] = headers[userAgent] || `npm (${this.userAgent})`;
+
+    if (!this.config.auth) {
+      return headers;
+    }
+
+    // copy headers to normalize keys
+    let copyHeaders = {};
+    Object.keys(headers).map((value) => {
```
Comment (juanpicado) [inline src/lib/storage/up-storage.js]: Two questions.

* Is this is necessary?
* Why not at the end?

File: src/lib/storage/up-storage.js [hunk 2]
```diff
@@ -227,6 +227,57 @@ class ProxyStorage {
     headers[acceptEncoding] = headers[acceptEncoding] || 'gzip';
     // registry.npmjs.org will only return search result if user-agent include string 'npm'
     headers[userAgent] = headers[userAgent] || `npm (${this.userAgent})`;
+
+    if (!this.config.auth) {
```
Comment (juanpicado) [inline src/lib/storage/up-storage.js]: This new block should be in a new private function, this is too big already.

File: src/lib/storage/up-storage.js [hunk 3]
```diff
@@ -227,6 +227,57 @@ class ProxyStorage {
...
+      return headers;
+    }
+
+    // copy headers to normalize keys
+    let copyHeaders = {};
```
Comment (juanpicado) [inline src/lib/storage/up-storage.js]: If you are not going to reassign, use `const` instead.

File: src/lib/storage/up-storage.js [hunk 4]
```diff
@@ -227,6 +227,57 @@ class ProxyStorage {
...
+    Object.keys(headers).map((value) => {
+      copyHeaders[value.toLowerCase()] = headers[value];
+    });
+
+    // if header Authorization assigns this has precedence
+    if (copyHeaders['authorization']) {
+      return headers;
+    }
+
+    if (typeof this.config.auth !== 'object') {
```
Comment (juanpicado) [inline src/lib/storage/up-storage.js]: Use lodash here, `_.isObject`

File: src/lib/storage/up-storage.js [hunk 5]
```diff
@@ -227,6 +227,57 @@ class ProxyStorage {
...
+    }
+
+    let token = null;
+    // define authorization token to use in the Authorization Header
+    if (this.config.auth.token) {
+      token = this.config.auth.token;
+    } else if (this.config.auth.token_env) {
+      // get NPM_TOKEN http://blog.npmjs.org/post/118393368555/deploying-with-npm-private-modules
+      // or get other variable export in env
+      if (this.config.auth.token_env === true) {
```
Comment (juanpicado) [inline src/lib/storage/up-storage.js]: === true ?

File: src/lib/storage/up-storage.js [hunk 6]
```diff
@@ -227,6 +227,57 @@ class ProxyStorage {
...
+
+    if (!token) {
+      throw new Error('Token is required');
+    }
+
+    // define type Auth allow basic and bearer
+    const type = this.config.auth.type;
+    switch (type.toLowerCase()) {
+      case 'basic':
+        headers['Authorization'] = `Basic ${token}`;
```
Comment (juanpicado) [inline src/lib/storage/up-storage.js]: Above 'Authorization' is lowercase and here is upper case.

File: src/lib/storage/up-storage.js [hunk 7]
```diff
@@ -227,6 +227,57 @@ class ProxyStorage {
...
+        break;
+      case 'bearer':
+        headers['Authorization'] = `Bearer ${token}`;
```
Comment (juanpicado) [inline src/lib/storage/up-storage.js]: You might create a private function to set this line, it is repeated above.

File: src/lib/storage/up-storage.js (hunk 1 above)
Comment (ramonornela) [reply src/lib/storage/up-storage.js]: This to normalize name headers

```
input: Authorization, authOrization, authoriZaTion

to: authorization

```

File: src/lib/storage/up-storage.js (hunk 5 above)
Comment (ramonornela) [reply src/lib/storage/up-storage.js]: To get value assigns in:
```shell
export NPM_TOKEN = "MY-jkdsjfksdjkfsdjfks"
```

Maybe is best, if assign auth.type get process.env.NPM_TOKEN and remove the value true.
```
if (!token) {
   token = process.env.NPM_TOKEN;
}
```

File: src/lib/storage/up-storage.js (hunk 1 above)
Comment (juanpicado) [reply src/lib/storage/up-storage.js]: But you could di it below, right?

Comment (ramonornela) [reply src/lib/storage/up-storage.js]: Yes, but removed on commit below https://github.com/verdaccio/verdaccio/pull/332/commits/ab69b258ca15f32b76290a19e2b2175ab63bf1fd

File: src/lib/storage/up-storage.js (hunk 4 above)
Comment (ramonornela) [reply src/lib/storage/up-storage.js]: Done

File: src/lib/storage/up-storage.js (hunk 5 above)
Comment (ramonornela) [reply src/lib/storage/up-storage.js]: Done

File: src/lib/storage/up-storage.js (hunk 6 above)
Comment (ramonornela) [reply src/lib/storage/up-storage.js]: Done

File: src/lib/storage/up-storage.js (hunk 7 above)
Comment (ramonornela) [reply src/lib/storage/up-storage.js]: Done

**Comment by ramonornela** [comment]
Related #124

File: test/functional/uplink.auth.spec.js [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+      };
+
+      assert.throws(fnError, 'Auth invalid');
+    });
+
+    it('if set headers authorization', function () {
+      const headers = setHeaders({}, {
+        'authorization': 'basic Zm9vX2Jhcg=='
+      });
+      assert.equal(Object.keys(headers).length, 4);
```
Comment (juanpicado) [inline test/functional/uplink.auth.spec.js]: Just to be consistent, you left break lines between assets and assignment in some test which is great. Could you do in the others as well?

File: src/lib/storage/up-storage.js [hunk 9]
```diff
@@ -227,10 +227,67 @@ class ProxyStorage {
...
+   * @private
+   */
+  _setAuth(headers) {
+
+    if (_.isUndefined(this.config.auth)) {
+      return headers;
+    }
+
+    // if header Authorization assigns this has precedence
+    if (headers['authorization']) {
```
Comment (juanpicado) [inline src/lib/storage/up-storage.js]: I think you could combine this return with the one above. At the end, they return the same.

File: src/lib/storage/up-storage.js [hunk 10]
```diff
@@ -227,10 +227,67 @@ class ProxyStorage {
...
+      return headers;
+    }
+
+    if (!_.isObject(this.config.auth)) {
+      throw new Error('Auth invalid');
```
Comment (juanpicado) [inline src/lib/storage/up-storage.js]: I think we should also log this part. I guess will crash but if it does not we need to notify users why has crashed.

**Comment by ramonornela** [comment]
@juanpicado adjusts done, thanks!

**Comment by juanpicado** [comment]
Perfect @ramonornela great contribution !! Close #124

**Pull request merged**